FAP.pm

```diff
@@ -3522,30 +3509,75 @@ sub make_position($$$$$$;$)
 			}
 			$retstring .= sprintf("%03d/%03d", $course, $speed);
 		}
+	}
 		
-		if (defined $altitude) {
-			$altitude = $altitude / $feet_to_meters;
-			# /A=(-\d{5}|\d{6})
-			if ($altitude >= 0) {
-				$retstring .= sprintf("/A=%06.0f", $altitude);
-			} else {
-				$retstring .= sprintf("/A=-%05.0f", $altitude * -1);
-			}
+	if (defined $altitude) {
+		$altitude = $altitude / $feet_to_meters;
+		# /A=(-\d{5}|\d{6})
+		if ($altitude >= 0) {
+			$retstring .= sprintf("/A=%06.0f", $altitude);
+		} else {
+			$retstring .= sprintf("/A=-%05.0f", $altitude * -1);
 		}
-		
-		if ($options->{'comment'}) {
-			$retstring .= $options->{'comment'};
+	}
+	
+	if ($options->{'comment'}) {
+		$retstring .= $options->{'comment'};
+	}
+	
+	if ($options->{'dao'}) {
+		# !DAO! extension, use Base91 format for best precision
+		# /1.1 : scale from 0.99 to 0..90 for base91, int(... + 0.5): round to nearest integer
+		my $dao = '!w' . chr(int($latmin_dao/1.1 + 0.5) + 33) . chr(int($lonmin_dao/1.1 + 0.5) + 33) . '!';
+		$retstring .= $dao;
+	}
+	
+	if (defined($options->{'timestamp'})) {
+		if (not($options->{'timestamp'} =~ /^\d+$/)) {
+			return undef;
 		}
-		
-		if ($options->{'dao'}) {
-			# !DAO! extension, use Base91 format for best precision
-			# /1.1 : scale from 0.99 to 0..90 for base91, int(... + 0.5): round to nearest integer
-			my $dao = '!w' . chr(int($latmin_dao/1.1 + 0.5) + 33) . chr(int($lonmin_dao/1.1 + 0.5) + 33) . '!';
-			$retstring .= $dao;
+		my $now = time();
+		if ($options->{'timestamp'} == 0) {
+			$options->{'timestamp'} = $now;
 		}
 		
-		return $retstring;
+		my $age = $now - $options->{'timestamp'};
+		
+		if ($age < -3610) {
+			# over 1 hour into the future, fail
+			return undef;
+		} elsif ($age < 86400-3600) {
+			# within the last 23 hours, HMS
+			$retstring = make_timestamp($options->{'timestamp'}, 1) . $retstring;
+		} elsif ($age < 28*86400) {
+			# within the last 28 days, DHM
+			$retstring = make_timestamp($options->{'timestamp'}, 0) . $retstring;
+		} else {
+			# over 28 days into the past, fail
+			return undef;
+		}
+		# check for failed timestamp generation
+		if (not(defined($retstring))) {
+			return undef;
+		}
 	}
+	
+	# add the correct packet type character based on messaging and timestamp
+	if (defined($options->{'timestamp'})) {
```

Comment:
I could be wrong, but it appears current version of make_position already has some issues that would cause issues when using the make_object sub.  However adding this in the current form would further the issues.
[Suggestion] - move the building out of the compressed/normal format, altitude, comment, etc into their own sub and point make_object to the new one.

Reply:
The following example of the  results in a '!' being added after the timestamp.  

```
use Ham::APRS::FAP qw(parseaprs);
use Ham::APRS::FAP qw(make_position);
use Ham::APRS::FAP qw(make_object);
use Date::Calc qw(check_date Today Date_to_Time Add_Delta_YM Mktime);
use JSON;

my $aprspacket = "OH2RDP>BEACON,OH2RDG*,WIDE:;LEADER   *092345z4903.50N/07201.75W>088/036";

my %packetdata;
my $retval = parseaprs($aprspacket, \%packetdata);

my $json = JSON->new->utf8->pretty->encode(\%packetdata);
print "$json\n";
print "$retval\n";

my $packet = make_object("LEADER"
	, 1670629500
	, 49.0583333333333
	, 72.0291666666667
	, "/>"
	, 66.672
	, 88
	, undef
	, 1
	, 0
	, 0
	, undef);

my $name = "LEADER";
my $alive = 1;
my $lat = 49.0583333333333;
my $lon = -72.0291666666667;
my $speed = 66.672;
my $course = 88;
my $altitude = undef;
my $sym = "/>";

my $ambiguity = 1;
my $comment = "Comment blah";
my $time = 1670629500;

my $packet = make_object($name, $time, $lat, $lon, $sym, $speed, $course, $altitude, $alive, 0, 0, undef);
```